### PR TITLE
Support non task data environment, step 1

### PR DESCRIPTION
**Pull request by jbirddog**
Work on #1567 

Starting to bring in some changes from the `non_task_data_based_env` branch that has been in progress for a while. Main changes are moving `CustomScriptEngineEnvironment` from a hardcoded choice to a factory method so the script engine environment can be selected at runtime. This allows running tests or deploying with the non task data based environment without code changes. Also added the ability to omit the non task data state when serializing, which is required to get the migrator tests to pass _edit: when using the non task data based environment_.

## Summary by CodeRabbit

- **New Features**
	- Added support for environment variables in the development setup with a new `env_file` configuration.

- **Enhancements**
	- Introduced a new configuration option for script engine environments.
	- Improved script engine environment management with new class structures and dynamic creation methods.

- **Bug Fixes**
	- Updated test cases to include a new argument for script engine state serialization.

- **Refactor**
	- Refactored script engine environment classes for better modularity and maintainability.

### CODE REVIEW

File: spiffworkflow-backend/src/spiffworkflow_backend/services/process_instance_processor.py [hunk 1]
```diff
@@ -203,7 +188,27 @@ def revise_state_with_task_data(self, task: SpiffTask) -> None:
...
+        context: dict[str, Any],
+        external_context: dict[str, Any] | None = None,
+    ) -> bool:
+        super().execute(script, context, external_context)
+        for key in self._non_user_defined_keys:
+            if key in context:
+                context.pop(key)
+        self._last_result = context
+        return True
+
```
[the remark this answers is not in the document]
Comment (jbirddog) [reply spiffworkflow-backend/src/spiffworkflow_backend/services/process_instance_processor.py:209]: The intent is to remove keys from `context` that are present in `self._non_user_defined_keys`.

File: spiffworkflow-backend/src/spiffworkflow_backend/services/process_instance_processor.py [hunk 2]
```diff
@@ -162,25 +162,7 @@ class MissingProcessInfoError(Exception):
...
-        external_context: dict[str, Any] | None = None,
-    ) -> bool:
-        super().execute(script, context, external_context)
-        for key in self._non_user_defined_keys:
-            if key in context:
-                context.pop(key)
-        self._last_result = context
-        return True
-
+class BaseCustomScriptEngineEnvironment(BasePythonScriptEngineEnvironment):  # type: ignore
```
[the remark this answers is not in the document]
Comment (jbirddog) [reply spiffworkflow-backend/src/spiffworkflow_backend/services/process_instance_processor.py:165]: This is mostly a copy/paste refactor to introduce a base class `BaseCustomScriptEngineEnvironment` so the new factory method can have a return type for mypy. The methods were extracted from `TaskDataBasedEnvironment` which is used heavily from all our existing tests. The implementations here are largely `pass` or `return {}` so i think its ok to assume if the existing tests pass the refactor is good.

File: spiffworkflow-backend/tests/spiffworkflow_backend/unit/test_process_instance_migrator.py [hunk 3]
```diff
@@ -111,7 +111,7 @@ def test_can_run_version_4_migration(
         processor = ProcessInstanceProcessor(
             process_instance, include_task_data_for_completed_tasks=True, include_completed_subprocesses=True
         )
-        bpmn_process_dict_version_4 = processor.serialize()
+        bpmn_process_dict_version_4 = processor.serialize(serialize_script_engine_state=False)
```
Comment (jasquat) [inline spiffworkflow-backend/tests/spiffworkflow_backend/unit/test_process_instance_migrator.py:114]: Tests are passing for me fine with default settings. Is it actually necessary? I don't think the current changes should require this.

Comment (jbirddog) [reply spiffworkflow-backend/tests/spiffworkflow_backend/unit/test_process_instance_migrator.py:114]: I edited the description, which upon a re-read wasn't very clear, but these changes are to get the migrator tests passing with the non task data based environment. This was part of the work on the `non_task_data_env` branch to get the unit tests passing.

**Review by burnettk — APPROVED** [review]

**Review by jasquat — APPROVED** [review]

**Comment by dinithihj** [comment]
@jbirddog @burnettk Is there anything specific we need to test related to this ticket?

**Comment by jbirddog** [comment]
> @jbirddog @burnettk Is there anything specific we need to test related to this ticket?

Nothing specific, should have no noticeable impact on the environments.

**Pull request merged**